### PR TITLE
Add yaml config file with configurable k8s knobs

### PR DESCRIPTION
**Pull request by rgrandl**

### CODE REVIEW

File: internal/impl/kube.go [hunk 1]
```diff
@@ -330,46 +331,45 @@ func buildContainer(d deployment, g group) (corev1.Container, error) {
...
-		probe := &corev1.Probe{}
-		if opts.TimeoutSecs > 0 {
-			probe.TimeoutSeconds = opts.TimeoutSecs
-		}
-		if opts.PeriodSecs > 0 {
-			probe.PeriodSeconds = opts.PeriodSecs
-		}
-		if opts.SuccessThreshold > 0 {
-			probe.SuccessThreshold = opts.SuccessThreshold
+	// Add volume mounts if any volume specified for the application.
```
Comment (mwhittaker) [inline internal/impl/kube.go:334]: Same as the comment above, can we combine this loop and the following loop?

File: internal/impl/kube.go [hunk 2]
```diff
@@ -766,7 +785,7 @@ func newDeployment(app *protos.AppConfig, cfg *kubeConfig, depId, image string)
 }
 
 // newListener returns a new listener.
-func newListener(depId string, config *kubeConfig, name string) listener {
+func newListener(depId string, config *kubeConfig, name string) (listener, error) {
```
Comment (mwhittaker) [inline internal/impl/kube.go:788]: I don't think this function ever returns a non-nil error. Maybe we can just return a `listener` without the `error`?

**Review by mwhittaker — COMMENTED** [review]
Great change Robert! Really like the yaml :)

File: internal/impl/config.go [hunk 3]
```diff
@@ -14,10 +14,17 @@
...
-// [kube] section of a weaver.toml file.
+import (
+	autoscalingv2 "k8s.io/api/autoscaling/v2"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// kubeConfig contains the kubernetes configuration for a Service Weaver
+// application deployed with `weaver kube deploy`.
 type kubeConfig struct {
+	// Path to the app config file.
```
[the remark this answers is not in the document]
Comment (rgrandl) [reply internal/impl/config.go:25]: Good point. Done

File: internal/impl/kube.go [hunk 4]
```diff
@@ -182,7 +153,30 @@ func buildDeployment(d deployment, g group) (*appsv1.Deployment, error) {
 				},
 			},
 		},
-	}, nil
+	}
+
+	// Add volume sources if any volume specified for the application.
+	for _, v := range d.config.Volumes {
```
[the remark this answers is not in the document]
Comment (rgrandl) [reply internal/impl/kube.go:159]: Interesting. Didn't know you can do that.

File: internal/impl/kube.go (hunk 2 above)
Comment (rgrandl) [reply internal/impl/kube.go:788]: Good catch

File: internal/impl/config.go [hunk 5]
```diff
@@ -51,70 +58,90 @@ type kubeConfig struct {
...
-	// "loki_service"       - to export logs to Grafana Loki [3]
-	// "grafana_service"    - to visualize/manipulate observability information [4]
+	// [1] https://pkg.go.dev/k8s.io/kubernetes/pkg/apis/autoscaling#HorizontalPodAutoscalerSpec.
+	//
+	// TODO(rgrandl): we might want to allow the user to specify the scaling spec
+	// differently for each group of collocated components.
+	ScalingSpec *autoscalingv2.HorizontalPodAutoscalerSpec
+
+	// Volumes that should be provided to all the running components.
+	Volumes []volume
```
[the remark this answers is not in the document]
Comment (rgrandl) [reply internal/impl/config.go:93]: Decoupling volumes from volume mounts sounds like a good idea. Done

**Review by rgrandl — COMMENTED** [review]
Thanks Michael for the thorough review.

File: internal/impl/config.go [hunk 6]
```diff
@@ -127,67 +154,38 @@ type listenerConfig struct {
...
-// httpAction describes the probe action when using HTTP. It mirrors Kubernetes
-// HTTPGetAction [1].
-//
-// [1] https://github.com/kubernetes/api/blob/v0.28.3/core/v1/types.go#L2208
-type httpAction struct {
-	Path string // Path to access on the HTTP server.
-	Port int32  // Port number to access on the container.
+// VolumeSpecs encapsulates volumes and volume mounts specs as defined by the
+// user in the kubernetes config.
+type VolumeSpecs struct {
```
Comment (mwhittaker) [inline internal/impl/config.go:173]: Can we unexport this?

**Review by mwhittaker — APPROVED** [review]
Nice!

**Pull request merged**